stripe/stripe-dotnet
# Fix deserialization logic for polymorphic types

r? @remi-stripe @talljoe-stripe 
cc @stripe/api-libraries 

Fixes the deserialization logic to properly handle polymorphic types, i.e. properties that are declared with an interface as their type rather than a concrete class.

Summary of changes:
- new `Stripe.Infrastructure.Util` class that contains:
    - a mapping of `object` values to concrete types, e.g. `"charge" => typeof(Stripe.Charge)`
    - a `GetConcreteType()` method that accepts a "potential type" and an `object` value:
        - if the potential type is a concrete type (not an interface), return it
        - otherwise, use the `object` value to find the matching concrete type, and return it (if it's compatible with the interface)
- `AbstractStripeObjectConverter`, `BalanceTransactionSourceConverter`, `ExternalAccountConverter`, `PaymentSourceConverter` and `StripeObjectConverter` are all gone
- new `StripeObjectConverter` deserializer (same name, but completely different logic) that uses `Util.GetConcreteType()` to figure out which concrete type to instantiate and populate
    - the converter is directly applied to properties with interface types (e.g. `EventData.Object`), which ensures that the class can be deserialized with `JsonConvert.DeserializeObject()`
    - the converter is also used as the `ItemConverterType` for `StripeList.Data` so that `StripeList` can be used with interfaces (e.g. `StripeList<IPaymentSource>` on `Customer`)
- `StringOrObject` class used for expandable fields also uses `Util.GetConcreteType()` to decide which type to pass to `JObject.ToObject()` for deserialization
- `Stripe.Mapper` now exposes `SerializerSettings` property that can be customized by users. Default settings are:
    - Use the `StripeObjectConverter` deserializer everywhere (which ensures that interfaces can always be deserialized)
    - Don't parse dates (which fixes #1178)

Test changes:
- Use `JsonConvert.DeserializeObject<>` instead of `Mapper<>.MapFromJson` wherever possible (i.e. when the type is not an interface)
- Add a new wholesome test to ensure that all model classes that implement `IHasObject` have an entry in the `Util.ObjectsToTypes` dictionary
- Add a new test to ensure that strings that contain ISO dates are not changed
- Add a test to ensure that serializing/deserializing/reserializing produces consistent results
